### PR TITLE
[ext] Initialise extension

### PR DESCRIPTION
**Pull request by zerefwayne**
**Description**

This PR initializes the Chrome helper Extension project. The project will use TypeScript and Webpack for transpilation purposes.

All the code for background service worker will reside in the `src` folder while the resources required for configuring the extension reside insie `public` folder.

The .ts files inside the `src` folder get transpiled and bundled into a single `build/index.js` file which is loaded by `manifest.json`. Once the build completes, contents of the  `public` folder get copied into `build` folder. This `build` directory can be loaded into chrome as an extension.

NOTE: `package-lock.json` contributes +10834 in the PR. Manually written code: +161 lines

**Screenshots**

image

### CODE REVIEW

File: diagnostics-extension/.eslintrc [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    "plugin:@typescript-eslint/recommended"
+  ],
+  "rules": {
+    "no-console": "off"
+  },
+  "env": {
+    "node": true,
+    "browser": true
+  }
+}
```
Comment (zerefwayne) [inline diagnostics-extension/.eslintrc:19]: Default ESLint Typescript configuration

File: diagnostics-extension/jest.config.js [hunk 2]
```diff
@@ -0,0 +1,6 @@
+/* eslint-disable */
+
+module.exports = {
+  preset: "ts-jest",
+  testEnvironment: "node",
+};
```
Comment (zerefwayne) [inline diagnostics-extension/jest.config.js:6]: Default config for testing the project using Jest library.

File: diagnostics-extension/package.json [hunk 3]
```diff
@@ -0,0 +1,40 @@
+{
+  "name": "diagnostics-extension",
+  "version": "1.0.0",
+  "description": "Extension for Sample Diagnostics App",
+  "main": "index.js",
+  "scripts": {
+    "prebuild": "sudo rm -rf build/",
+    "build": "webpack",
+    "postbuild": "cp -R public/** build",
```
Comment (zerefwayne) [inline diagnostics-extension/package.json:8]: When `npm run build` is run, the `pre` and `post` automatically run before and after the build command respectively.

File: diagnostics-extension/public/manifest.json [hunk 4]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "ChromeOS Diagnostics",
```
Comment (MahmoudAGawad) [inline diagnostics-extension/public/manifest.json]: nit: "Chrome OS Diagnostics Helper Extension"

File: diagnostics-extension/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+// Copyright 2021 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+/**
+ * @fileoverview Entry-point script of background worker
```
Comment (MahmoudAGawad) [inline diagnostics-extension/src/index.ts]: nit: @fileoverview Service worker script

File: diagnostics-extension/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+ */
+
+console.log('Background worker is up!');
```
Comment (MahmoudAGawad) [inline diagnostics-extension/src/index.ts]: remove this.

You should listen on install events and do the necessary pre-work (which will be empty for now).
```

self.addEventListener('install', function(event) {
   console.log('Service worker is installed!');
  );
});
```

**Review by MahmoudAGawad — APPROVED** [review]
LGTM. Thanks!

File: diagnostics-extension/public/manifest.json [hunk 7]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "Chrome OS Diagnostics Helper Extension",
```
Comment (Lamzin) [inline diagnostics-extension/public/manifest.json]: nit: let's maybe call it a `companion extension`?

File: diagnostics-extension/src/index.ts [hunk 8]
```diff
@@ -0,0 +1,13 @@
+// Copyright 2021 The Chromium Authors. All rights reserved.
```
Comment (Lamzin) [inline diagnostics-extension/src/index.ts]: nit: let's rename it to `sw.ts` to avoid confusion :)

**Review by Lamzin — APPROVED** [review]
LGTM, but can you please add build command to commit hooks and github CI?

**Comment by zerefwayne** [comment]
> LGTM, but can you please add build command to commit hooks and github CI?

@Lamzin Currently I'm not pushing the build folder to GitHub. Should I remove it from `.gitignore` => create a build after every commit => push it to github?

File: diagnostics-extension/webpack.config.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    rules: [
+      {
+        test: /\.ts$/,
+        use: "ts-loader",
+        include: [path.resolve(__dirname, "src")],
+      },
+    ],
+  },
+  output: {
+    filename: "index.js",
```
Comment (MahmoudAGawad) [inline diagnostics-extension/webpack.config.js]: sw.js?

File: diagnostics-extension/package.json [hunk 10]
```diff
@@ -0,0 +1,40 @@
+{
+  "name": "diagnostics-extension",
+  "version": "1.0.0",
+  "description": "Extension for Sample Diagnostics App",
+  "main": "index.js",
```
Comment (MahmoudAGawad) [inline diagnostics-extension/package.json]: sw.js?

File: diagnostics-extension/public/manifest.json [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+  "manifest_version": 3,
+  "icons": {
+    "16": "images/favicon-16x16.png",
+    "32": "images/favicon-32x32.png",
+    "48": "images/favicon-48x48.png",
+    "128": "images/favicon-120x120.png",
+    "192": "images/favicon-192x192.png"
+  },
+  "background": {
+    "service_worker": "index.js"
```
Comment (MahmoudAGawad) [inline diagnostics-extension/public/manifest.json]: sw.js

**Comment by Lamzin** [comment]
> @Lamzin Currently I'm not pushing the build folder to GitHub. Should I remove it from .gitignore => create a build after every commit => push it to github?

It looks like there is no automation for Chrome extension as we have for PWA, i.e. pre commit hook + github CI (.github/workflows/angular.test.yml).

**Comment by zerefwayne** [comment]
> > @Lamzin Currently I'm not pushing the build folder to GitHub. Should I remove it from .gitignore => create a build after every commit => push it to github?
> 
> It looks like there is no automation for Chrome extension as we have for PWA, i.e. pre commit hook + github CI (.github/workflows/angular.test.yml).

I can add a command to run the tests for extension. Should I add something for building it as well? I can check if the build was completed successfully for now. 

Update: @Lamzin I've added a similar set up and Github Action for extension.

>    paths:
      - "diagnostics-extension/**"

The extension CI will run when we make some changes in the extension directory.

I can add a similar paths variable in the pwa one but for that I plan to make a new ( PR #22 ) since it's out of scope of this one.

File: diagnostics-extension/README.md [hunk 12]
```diff
@@ -0,0 +1,21 @@
+# ChromeOS Diagnotics Helper Extension
```
Comment (Lamzin) [inline diagnostics-extension/README.md]: nit: companion extension

**Review by Lamzin — APPROVED** [review]
LGTM % nit, thanks a lot!

**Review by MahmoudAGawad — APPROVED** [review]
Still LGTM. Thanks!

**Pull request merged**